README.md
<div align="center"><a href="https://github.com/AutohomeCorp/frostmourne"><img src="https://gitee.com/tim_guai/frostmourne/raw/master/doc/img/logo/frostmourne_logo.jpg"></a></div>

<p align="center">
<a href="https://github.com/AutohomeCorp/frostmourne/releases"><img src="https://img.shields.io/github/release/AutohomeCorp/frostmourne" alt="GitHub release"></a>
<a href="https://github.com/AutohomeCorp/frostmourne/stargazers"><img src="https://img.shields.io/github/stars/AutohomeCorp/frostmourne" alt="GitHub Stars"></a>
<a href="https://github.com/AutohomeCorp/frostmourne/fork"><img src="https://img.shields.io/github/forks/AutohomeCorp/frostmourne" alt="GitHub Forks"></a>
<a href="https://github.com/AutohomeCorp/frostmourne/graphs/contributors"><img src="https://img.shields.io/github/contributors/AutohomeCorp/frostmourne" alt="GitHub Contributors"></a>
<a href="https://github.com/AutohomeCorp/frostmourne/issues"><img src="https://img.shields.io/github/issues/AutohomeCorp/frostmourne" alt="GitHub issues"></a>
<a href="https://openjdk.java.net/"><img src="https://img.shields.io/badge/Java-8,11,13-blue?logo=java&logoColor=white" alt="JDK support"></a>
<a href="https://github.com/AutohomeCorp/frostmourne/blob/master/LICENSE"><img src="https://img.shields.io/npm/l/svelte.svg" alt="LICENSE"></a>
<a href="https://github.com/AutohomeCorp/frostmourne"><img src="https://img.shields.io/github/downloads/AutohomeCorp/frostmourne/total.svg" alt="Downloads"></a>
</p>

[Github地址](https://github.com/AutohomeCorp/frostmourne) | [Gitee地址](https://gitee.com/tim_guai/frostmourne) | [中文](./README.md) | [English](./README_en.md)


# 📖 介绍

`Frostmourne`(霜之哀伤)是汽车之家经销商技术部监控系统的开源版本，用于帮助监控几乎所有数据库数据(包括`Elasticsearch`, `Prometheus`, `SkyWalking`, `MySql` 等等)。如果你已经建立起了日志系统，
指标体系，却苦恼于没有一个配套监控系统，也许它能帮到你。使用本系统得当，至少可以做到线上应用出了什么问题，相关开发能第一时间了解报错情况，知道问题出在哪里，
可以快速响应处理，几乎先于所有人知道线上的情况，也因为心中有数所以可以从容应对各种问询，一点都不慌，你们团队值得拥有。

### 项目初衷

在用`ELK`建立起日志系统之后，我们发现应用日志监控这块除了`ElastAlert`之外，没有其他方案。我们初期使用`ElastAlert`来解决日志监控的问题，
但是随着配置的增加，不仅管理成本和使用成本较高，稳定性方面也不能让我们满意，所以为了更好的易用性，稳定性，我们决定自己做一套简单的监控系统，
来解决日志监控的问题。

项目并不局限于`Elasticsearch`数据，还支持`HTTP`, `PING`, `Telnet`监控，`Prometheus`, `SkyWalking`, `InfluxDB`，`MySQL`, `ClickHouse`, `IoTDB`, `SqlServer`数据监控，
后面还会加入更多的常用数据源(如：`Loki`, `MongoDB`, `Redis`, `Oracle` 等)纳入监控范畴，需要做的东西还有很多，需要更多相关开发加入进来，欢迎联系我们，一起做大做强。

# ✨ 主要功能

* 只需要写一条数据查询就可以轻松搞定监控
* 多种数据源支持：`Elasticsearch, HTTP, SkyWalking, Prometheus, InfluxDB, MySQL/TiDb, ClickHouse, SqlServer, PING, IotDB, Telnet`
* 数值计算类型监控：`count, min, max, avg, sum, unique count, percentiles, standard deviation`; `Elasticsearch`数据支持分桶
* 多种报警消息发送方式：钉钉(机器人)、企业微信(机器人)、飞书机器人、Email、短信、HTTP
* 多种消息格式：`text, markdown`
* 灵活的报警消息`Freemarker`模板定制，支持变量占位符；消息模板管理
* 分布式调度实现，每个监控都是独立调度，互不影响
* 报警消息附带日志查询短链接，直达报警原因
* 数值同比，环比监控
* `HTTP`数据监控, `Javascript`表达式判断是否报警; `PING`连通监控, `Telnet`端口连通监控
* 前端简单易用：监控管理、测试、另存、执行日志和历史消息
* `Elasticsearch`数据查询、分享和下载
* 报警消息抑制功能，防止消息轰炸；也有报警升级功能，避免故障相关方长时间得不到通知。
* 自带账号，团队，部门信息管理模块，也可自己实现内部对接
* 集成`LDAP`登录认证
* 权限控制，数据隔离，各团队互不影响

    #### 企业微信机器人报警截图展示： （<a href="./doc/wiki/feature_image.md" target="_blank">查看完整功能截图展示</a>）

    <img src="https://gitee.com/tim_guai/frostmourne/raw/master/doc/wiki/img/markdown_wechat_robot.png" />

<!-- TOC -->

## 目录

- [介绍](#-介绍)
- [主要功能](#-主要功能)
- [在线demo](#-在线demo)
- [快速启动](#-快速启动)
- [部署](#-部署)
  - [环境准备](#预备环境准备mysql数据库表创建)
  - [k8s部署方式](#一k8s部署方式)
  - [zip包部署方式](#二zip包部署方式)
  - [自构建部署方式](#三自构建部署方式)
- [功能使用指南](#-功能使用指南)
- [开发调试](#-开发调试)
- [后续规划](#后续规划)
- [发版历史](./ReleaseNotes.md)
- [Contributors](#-contributors)

<!-- /TOC -->

# 💻 在线demo

为了更快的理解本项目的作用，提供了一个接口全`mock`的静态站点供大家预览功能: <a href="https://frostmourne-demo.github.io/">在线demo</a>
在线`demo`更新不及时，请以项目实际运行效果为准，`demo`只是用于快速浏览


# 📦 快速启动

提供`docker-compose`方式，让你更快运行起来便于更好理解项目作用。详细请看文档：<a href="./doc/wiki/quick-start.md" target="_blank">Quick-Start</a>


# 🧰 部署

#### 预备环境准备：`MySQL`数据库表创建
> Frostmourne所有表的创建语句都在 [frostmourne.sql](./doc/mysql-schema/frostmourne.sql) 文件中。

数据库密码默认使用明文，没有加密策略。如果你需要对密码进行加密，请参考druid官方文档：[druid数据库密码加密](https://github.com/alibaba/druid/wiki/%E4%BD%BF%E7%94%A8ConfigFilter)


#### 一、`k8s`部署方式
`k8s`部署参考以下三个配置文件

* [frostmourne-monitor-namespace.yaml](./doc/docker/k8s/frostmourne-monitor-namespace.yaml)
* [frostmourne-monitor-deployment.yaml](./doc/docker/k8s/frostmourne-monitor-deployment.yaml)
* [frostmourne-monitor-service.yaml](./doc/docker/k8s/frostmourne-monitor-service.yaml)

相关参数在 `frostmourne-monitor-deployment.yaml` 文件里配置。需要注意的是在`frostmourne-monitor-service.yaml`里指定对外映射端口，默认`nodePort=30054`

```bash
kubectl applt -f frostmourne-monitor-namespace.yaml
kubectl applt -f frostmourne-monitor-deployment.yaml
kubectl apply -f frostmourne-monitor-service.yaml
```

#### 二、`zip`包部署方式

依赖环境
* `JDK 1.8`
* `MySQL 5.7.8+`

最新的Release版本zip包，请到 [releases](https://github.com/AutohomeCorp/frostmourne/releases) 中下载，解压后然后根据自己的环境修改应用配置文件`application.properties`文件和环境变量配置文件`env`，然后执行如下命令启动：

```bash
./scripts/startup.sh
```

执行如下命令停止应用：

```bash
./scripts/shutdown.sh
```

#### 三、自构建部署方式

依赖环境
* `JDK 1.8`
* `Maven 3.2.x+`
* `MySQL 5.7.8+`

在项目`frostmourne`主目录下执`maven`构建命令：
```bash
mvn -U clean package -DskipTests=true
```

前端项目`frostmourne-vue`会自动把资源构建到`frostmourne-monitor`的`resources/dist`下，所以你只需要部署`frostmourne-monitor`即可。

`frostmourne-monitor`已经配置了`assembly`打包，`target`目录下会生成`zip`包，你只需要将`zip`包解压，然后根据自己的
环境修改应用配置文件`application.properties`文件和环境变量配置文件env，然后执行如下命令启动：

```bash
./scripts/startup.sh
```

执行如下命令停止应用：

```bash
./scripts/shutdown.sh
```

# 📚 功能使用指南

<table>
<tr>
<td><a href="./doc/wiki/es.md" target="_blank">Elasticsearch数据监控指南</a></td>
<td><a href="./doc/wiki/http-alarm.md" target="_blank">HTTP监控使用说明</a></td>
<td><a href="./doc/wiki/prometheus.md" target="_blank">Prometheus数据监控指南</a></td>
</tr>
<tr>
<td><a href="./doc/wiki/skywalking.md" target="_blank">SkyWalking数据监控指南</a></td>
<td><a href="./doc/wiki/jdbc-mysql.md" target="_blank">MySQL/TiDB数据监控指南</a></td>
<td><a href="./doc/wiki/jdbc-clickhouse.md" target="_blank">ClickHouse数据监控指南</a></td>
</tr>
<tr>

ReleaseNotes.md
# 1.1-SNAPSHOT

### Feature

* 【1.1】发布1.0-RELEASE

# 1.0-RELEASE

### Feature

* 【1.0】发布0.9-RELEASE, 开始1.0-SNAPSHOT开发 [2022-09-19]
* 【1.0】数据名增加名称模糊查询 [2022-09-25]
* 【1.0】prometheus增加/api/v1/query_range查询支持 [2022-10-08]
* 【1.0】增加监控转团队功能 [2022-11-19]
* 【1.0】修复团队名称修改时，监控和组对不上的问题 [2022-11-19]
* 【1.0】ldap用户验证通过，系统还不存在该用户时自动创建 [2023-01-02]
* 【1.0】featrue:报警方式增加OneMessage机器人 - [SQL](./doc/mysql-schema/2023-02-24/change.sql) [2023-02-24]

### BugFix

* 【1.0】bugfix: 修复es8数据查询排序不生效的问题 [2023-02-23]
* 【1.0】bugfix: 修复监控执行日志是否报警查询条件不生效的问题 [2023-02-23]
* 【1.0】bugfix: 解决es8数据下载报错的问题 [2023-03-16]

### Mysql

* 【1.0】表alert, alert_upgrade增加字段one_message_robot_hook - [SQL](./doc/mysql-schema/2023-02-24/change.sql) [2023-02-24]

### Other

* 【1.0】启动脚本优化，多版本jar包使用最新版本启动 [2022-09-23]

# 0.9.1-RELEASE

### Bugfix

* 【0.9.1】修复 http, telent 等无数据源的类型，获取 数据源 settings 报空指针问题 [2022-09-22]

# 0.9-RELEASE

### Feature

* 【0.9】发布0.8-RELEASE, 进入0.9开发 [2022-06-09]
* 【0.9】增加telnet端口连通监控 [2022-06-09]
* 【0.9】增加SqlServer数据监控报警 [2022-06-30]
* 【0.9】监控列表增加监控调度时间查看 [2022-08-04]
* 【0.9】优化启动脚本，支持java11, java13 [2022-08-04]
* 【0.9】告警列表组件新增日志查询 [2022-08-13]
* 【0.9】数据源及数据名页面数据类型新增icon [2022-08-13]
* 【0.9】后端接口增加部分数据校验并返回合适的提示信息 [2022-08-17]
* 【0.9】启动增加时区参数TZ配置 [2022-08-19]
* 【0.9】敏感信息加密处理 [2022-09-19]

### BugFix

* 【0.9】修复飞书消息发送两条的问题 [2022-06-17]

### Document

* 【0.9】增加telnet端口监控使用指南 [telnet.md](./doc/wiki/telnet.md) [2022-06-09]

### Thanks

0.9-RELEASE大部分功能由 [@limbo](https://github.com/limboinf) 贡献，感谢！

# 0.8-RELEASE

### Feature

* 【0.8】 发布0.7.1-RELEASE, 进入0.8-SNAPSHOT开发版 [2022-05-14]
* 【0.8】邮箱报警支持ssl [2022-05-15]
* 【0.8】增加报警升级功能 [2022-05-15]
* 【0.8】增加国际化支持 [2022-05-28]
* 【0.8】mysql, clickhouse表达式规则增加TOP 50条记录数据TOP_N_DOCUMENTS [2022-05-31]
* 【0.8】Elasticsearch数值实现环比监控 [2022-06-07]
* 【0.8】influxdb数值实现环比监控 [2022-06-09]
* 【0.8】增加 [iotdb](https://github.com/apache/iotdb) 数据监控报警 [2022-06-09]

### BugFix

* 【0.8】解决logo不显示的问题 [2022-05-17]
* 【0.8】解决关闭数据库连接错误调用clone方法的问题 [gitee#I58JI1](https://gitee.com/tim_guai/frostmourne/issues/I58JI1) [2022-05-21]
* 【0.8】优化druid连接参数，修复连接超时后无法执行sql的问题 [2022-05-27]
* 【0.8】解决消息模板列表模板类型不显示的问题 [2022-06-07]

### Mysql

* 【0.8】alert_event表新增upgrade字段; recipient新增recipient字段; 新增alert_upgrade表 - [SQL](./doc/mysql-schema/2022-05-15/change.sql) [2022-05-15]

### Others

* 【0.8】frostmourne-core改名为frostmourne-common [2022-05-18]
* 【0.8】metric注入逻辑重构，去掉metric map构造改成自动注入 [2022-06-09]

# 0.7.1-RELEASE

### Feature

* 发布0.6.2-RELEASE [2022-05-05]
* 【0.7】静默优化，支持字段判断表达式 [2022-05-05]
* 【0.7】mysql, clickhouse监控增加表达式监控规则 [2022-05-06]
* 【0.7】增加ping监控报警 [2022-05-07]
* 【0.7】增加[skywalking](https://github.com/apache/skywalking)日志数据监控 [2022-05-09]
* 【0.7】增加[skywalking](https://github.com/apache/skywalking)报警数据监控 [2022-05-10]
* 【0.7】增加[prometheus](https://github.com/prometheus/prometheus)数据监控报警支持 [2022-05-12]
* 【0.7.1】增加判断规则：持续n次满足规则时报警 [2022-05-13]

### Bugfix:

* 【0.7】bugfix: 解决http消息模板不加载的问题 [2022-05-12]

### Mysql


frostmourne-vue/mock/log.js
export default [
  {
    url: '/log/findAlarmLog',
    type: '',
    response: _ => {
      return {
        'returncode': 0,
        'message': 'ok',
        'result': {
          'rowcount': 938,
          'pagecount': 94,
          'pageindex': 1,
          'list': [
            {
              'id': 2395,
              'alarm_id': 22,
              'exe_start': '2020-01-13T19:02:05.000+0800',
              'exe_end': '2020-01-13T19:02:05.000+0800',
              'cost': 28,
              'execute_result': 'SUCCESS',
              'verify_result': 'FALSE',
              'create_at': '2020-01-13T19:02:05.000+0800',
              'message': '[2020-01-13 19:02:05] execute start\r\n[2020-01-13 19:02:05] start pull metric\r\n[2020-01-13 19:02:05] env = {"number_of_pending_tasks":0,"cluster_name":"es1","active_shards":7704,"active_primary_shards":3852,"unassigned_shards":0,"delayed_unassigned_shards":0,"HTTP_STATUS":200,"timed_out":false,"HTTP_COST":27,"relocating_shards":0,"initializing_shards":0,"task_max_waiting_in_queue_millis":0,"number_of_data_nodes":10,"number_of_in_flight_fetch":0,"active_shards_percent_as_number":100.0,"EXPRESSION":"status == \\"yellow\\" || status == \\"red\\" || number_of_nodes != 10","status":"green","number_of_nodes":10}\r\n[2020-01-13 19:02:05] execute end'
            },
            {
              'id': 2394,
              'alarm_id': 19,
              'exe_start': '2020-01-13T19:00:38.000+0800',
              'exe_end': '2020-01-13T19:00:38.000+0800',
              'cost': 171,
              'execute_result': 'SUCCESS',
              'verify_result': 'FALSE',
              'create_at': '2020-01-13T19:00:38.000+0800',
              'message': '[2020-01-13 19:00:37] execute start\r\n[2020-01-13 19:00:37] start pull metric\r\n[2020-01-13 19:00:37] env = {"NUMBER":0,"startTime":"2020-01-13T18:55:37.699+08:00","THRESHOLD":"1","endTime":"2020-01-13T19:00:37.699+08:00","TIME_WINDOW":"5","OPERATOR":"GTE"}\r\n[2020-01-13 19:00:37] execute end'
            },
            {
              'id': 2393,
              'alarm_id': 22,
              'exe_start': '2020-01-13T19:00:05.000+0800',
              'exe_end': '2020-01-13T19:00:05.000+0800',
              'cost': 33,
              'execute_result': 'SUCCESS',
              'verify_result': 'FALSE',
              'create_at': '2020-01-13T19:00:05.000+0800',
              'message': '[2020-01-13 19:00:05] execute start\r\n[2020-01-13 19:00:05] start pull metric\r\n[2020-01-13 19:00:05] env = {"number_of_pending_tasks":0,"cluster_name":"es1","active_shards":7704,"active_primary_shards":3852,"unassigned_shards":0,"delayed_unassigned_shards":0,"HTTP_STATUS":200,"timed_out":false,"HTTP_COST":30,"relocating_shards":0,"initializing_shards":0,"task_max_waiting_in_queue_millis":0,"number_of_data_nodes":10,"number_of_in_flight_fetch":0,"active_shards_percent_as_number":100.0,"EXPRESSION":"status == \\"yellow\\" || status == \\"red\\" || number_of_nodes != 10","status":"green","number_of_nodes":10}\r\n[2020-01-13 19:00:05] execute end'
            },
            {
              'id': 2392,
              'alarm_id': 22,
              'exe_start': '2020-01-13T18:58:05.000+0800',
              'exe_end': '2020-01-13T18:58:05.000+0800',
              'cost': 30,
              'execute_result': 'SUCCESS',
              'verify_result': 'FALSE',
              'create_at': '2020-01-13T18:58:05.000+0800',
              'message': '[2020-01-13 18:58:05] execute start\r\n[2020-01-13 18:58:05] start pull metric\r\n[2020-01-13 18:58:05] env = {"number_of_pending_tasks":0,"cluster_name":"es1","active_shards":7704,"active_primary_shards":3852,"unassigned_shards":0,"delayed_unassigned_shards":0,"HTTP_STATUS":200,"timed_out":false,"HTTP_COST":28,"relocating_shards":0,"initializing_shards":0,"task_max_waiting_in_queue_millis":0,"number_of_data_nodes":10,"number_of_in_flight_fetch":0,"active_shards_percent_as_number":100.0,"EXPRESSION":"status == \\"yellow\\" || status == \\"red\\" || number_of_nodes != 10","status":"green","number_of_nodes":10}\r\n[2020-01-13 18:58:05] execute end'
            },
            {
              'id': 2391,
              'alarm_id': 22,
              'exe_start': '2020-01-13T18:56:05.000+0800',
              'exe_end': '2020-01-13T18:56:05.000+0800',
              'cost': 44,
              'execute_result': 'SUCCESS',
              'verify_result': 'FALSE',
              'create_at': '2020-01-13T18:56:05.000+0800',
              'message': '[2020-01-13 18:56:05] execute start\r\n[2020-01-13 18:56:05] start pull metric\r\n[2020-01-13 18:56:05] env = {"number_of_pending_tasks":0,"cluster_name":"es1","active_shards":7704,"active_primary_shards":3852,"unassigned_shards":0,"delayed_unassigned_shards":0,"HTTP_STATUS":200,"timed_out":false,"HTTP_COST":43,"relocating_shards":0,"initializing_shards":0,"task_max_waiting_in_queue_millis":0,"number_of_data_nodes":10,"number_of_in_flight_fetch":0,"active_shards_percent_as_number":100.0,"EXPRESSION":"status == \\"yellow\\" || status == \\"red\\" || number_of_nodes != 10","status":"green","number_of_nodes":10}\r\n[2020-01-13 18:56:05] execute end'
            },
            {
              'id': 2390,
              'alarm_id': 19,
              'exe_start': '2020-01-13T18:55:35.000+0800',
              'exe_end': '2020-01-13T18:55:35.000+0800',
              'cost': 233,
              'execute_result': 'SUCCESS',
              'verify_result': 'FALSE',
              'create_at': '2020-01-13T18:55:35.000+0800',
              'message': '[2020-01-13 18:55:35] execute start\n[2020-01-13 18:55:35] start pull metric\n[2020-01-13 18:55:35] env = {"NUMBER":0,"startTime":"2020-01-13T18:50:35.020+08:00","THRESHOLD":"1","endTime":"2020-01-13T18:55:35.020+08:00","TIME_WINDOW":"5","OPERATOR":"GTE"}\n[2020-01-13 18:55:35] execute end'
            },
            {
              'id': 2389,
              'alarm_id': 22,
              'exe_start': '2020-01-13T18:54:05.000+0800',
              'exe_end': '2020-01-13T18:54:05.000+0800',
              'cost': 31,
              'execute_result': 'SUCCESS',
              'verify_result': 'FALSE',
              'create_at': '2020-01-13T18:54:05.000+0800',
              'message': '[2020-01-13 18:54:05] execute start\r\n[2020-01-13 18:54:05] start pull metric\r\n[2020-01-13 18:54:05] env = {"number_of_pending_tasks":0,"cluster_name":"es1","active_shards":7704,"active_primary_shards":3852,"unassigned_shards":0,"delayed_unassigned_shards":0,"HTTP_STATUS":200,"timed_out":false,"HTTP_COST":29,"relocating_shards":0,"initializing_shards":0,"task_max_waiting_in_queue_millis":0,"number_of_data_nodes":10,"number_of_in_flight_fetch":0,"active_shards_percent_as_number":100.0,"EXPRESSION":"status == \\"yellow\\" || status == \\"red\\" || number_of_nodes != 10","status":"green","number_of_nodes":10}\r\n[2020-01-13 18:54:05] execute end'
            },
            {
              'id': 2388,
              'alarm_id': 22,
              'exe_start': '2020-01-13T18:52:05.000+0800',
              'exe_end': '2020-01-13T18:52:05.000+0800',
              'cost': 39,
              'execute_result': 'SUCCESS',
              'verify_result': 'FALSE',
              'create_at': '2020-01-13T18:52:05.000+0800',
              'message': '[2020-01-13 18:52:05] execute start\n[2020-01-13 18:52:05] start pull metric\n[2020-01-13 18:52:05] env = {"number_of_pending_tasks":0,"cluster_name":"es1","active_shards":7704,"active_primary_shards":3852,"unassigned_shards":0,"delayed_unassigned_shards":0,"HTTP_STATUS":200,"timed_out":false,"HTTP_COST":25,"relocating_shards":0,"initializing_shards":0,"task_max_waiting_in_queue_millis":0,"number_of_data_nodes":10,"number_of_in_flight_fetch":0,"active_shards_percent_as_number":100.0,"EXPRESSION":"status == \\"yellow\\" || status == \\"red\\" || number_of_nodes != 10","status":"green","number_of_nodes":10}\n[2020-01-13 18:52:05] execute end'
            },
            {
              'id': 2387,
              'alarm_id': 19,
              'exe_start': '2020-01-13T18:51:04.000+0800',
              'exe_end': '2020-01-13T18:51:05.000+0800',
              'cost': 392,
              'execute_result': 'SUCCESS',
              'verify_result': 'FALSE',
              'create_at': '2020-01-13T18:51:05.000+0800',
              'message': '[2020-01-13 18:51:04] execute start\r\n[2020-01-13 18:51:04] start pull metric\r\n[2020-01-13 18:51:04] env = {"NUMBER":0,"startTime":"2020-01-13T18:46:04.397+08:00","THRESHOLD":"1","endTime":"2020-01-13T18:51:04.397+08:00","TIME_WINDOW":"5","OPERATOR":"GTE"}\r\n[2020-01-13 18:51:04] execute end'
            },
            {
              'id': 2386,
              'alarm_id': 22,
              'exe_start': '2020-01-13T18:50:05.000+0800',
              'exe_end': '2020-01-13T18:50:05.000+0800',
              'cost': 67,
              'execute_result': 'SUCCESS',
              'verify_result': 'FALSE',
              'create_at': '2020-01-13T18:50:05.000+0800',
              'message': '[2020-01-13 18:50:05] execute start\n[2020-01-13 18:50:05] start pull metric\n[2020-01-13 18:50:05] env = {"number_of_pending_tasks":0,"cluster_name":"es1","active_shards":7704,"active_primary_shards":3852,"unassigned_shards":0,"delayed_unassigned_shards":0,"HTTP_STATUS":200,"timed_out":false,"HTTP_COST":25,"relocating_shards":0,"initializing_shards":0,"task_max_waiting_in_queue_millis":0,"number_of_data_nodes":10,"number_of_in_flight_fetch":0,"active_shards_percent_as_number":100.0,"EXPRESSION":"status == \\"yellow\\" || status == \\"red\\" || number_of_nodes != 10","status":"green","number_of_nodes":10}\n[2020-01-13 18:50:05] execute end'
            }
          ]
        }
      }
    }
  },

  {
    url: '/log/findAlertLog',
    type: '',
    response: _ => {
      return {
        'returncode': 0,
        'message': 'ok',
        'result': {
          'rowcount': 31,
          'pagecount': 4,
          'pageindex': 1,
          'list': [
            {
              'id': 315,
              'alarm_id': 20,
              'execute_id': 2050,
              'way': 'dingding',
              'recipient': 'kechangqing',
              'in_silence': 'NO',
              'send_status': 'SUCCESS',
              'create_at': '2020-01-13T10:50:36.000+0800',
              'alert_type': 'RECOVER',
              'content': '消息类型: [恢复] 请自己检查问题是否解决,上次报警内容如下\n[2020-01-13 10:45:38]\napplication1最近5分钟内有异常日志1条。最近一条异常信息:\n服务器IP: 127.0.0.1\n异常类型: redis.clients.jedis.exceptions.JedisConnectionException\n自定义信息: HashCache error \n异常信息: save(key:xxxxxxxxxx, seconds:1200)\n\n详细请看: http://iii50.cn/uMiahM'
            },
            {
              'id': 316,
              'alarm_id': 20,
              'execute_id': 2050,
              'way': 'email',
              'recipient': 'kechangqing',
              'in_silence': 'NO',
              'send_status': 'SUCCESS',
              'create_at': '2020-01-13T10:50:36.000+0800',
              'alert_type': 'RECOVER',
              'content': '消息类型: [恢复] 请自己检查问题是否解决,上次报警内容如下\n[2020-01-13 10:45:38]\napplication1最近5分钟内有异常日志1条。最近一条异常信息:\n服务器IP: 127.0.0.1\n异常类型: redis.clients.jedis.exceptions.JedisConnectionException\n自定义信息: HashCache error \n异常信息: save(key:xxxxxxxx, seconds:1200)\n\n详细请看: http://iii50.cn/uMiahM'
            },
            {
              'id': 313,
              'alarm_id': 20,
              'execute_id': 2045,
              'way': 'dingding',
              'recipient': 'kechangqing',
              'in_silence': 'NO',
              'send_status': 'SUCCESS',
              'create_at': '2020-01-13T10:45:39.000+0800',
              'alert_type': 'PROBLEM',
              'content': '[2020-01-13 10:45:38]\napplication1最近5分钟内有异常日志1条。最近一条异常信息:\n服务器IP: 127.0.0.1\n异常类型: redis.clients.jedis.exceptions.JedisConnectionException\n自定义信息: HashCache error \n异常信息: save(key:xxxxxxx, seconds:1200)\n\n详细请看: http://iii50.cn/uMiahM'
            },
            {
              'id': 314,
              'alarm_id': 20,
              'execute_id': 2045,
              'way': 'email',
              'recipient': 'kechangqing',
              'in_silence': 'NO',
              'send_status': 'SUCCESS',
              'create_at': '2020-01-13T10:45:39.000+0800',
              'alert_type': 'PROBLEM',
              'content': '[2020-01-13 10:45:38]\napplication1最近5分钟内有异常日志1条。最近一条异常信息:\n服务器IP: 127.0.0.1\n异常类型: redis.clients.jedis.exceptions.JedisConnectionException\n自定义信息: HashCache error \n异常信息: save(key:xxxxxx, seconds:1200)\n\n详细请看: http://iii50.cn/uMiahM'
            },
            {
              'id': 311,
              'alarm_id': 20,
              'execute_id': 1954,
              'way': 'dingding',
              'recipient': 'kechangqing',
              'in_silence': 'NO',
              'send_status': 'SUCCESS',
              'create_at': '2020-01-13T09:10:35.000+0800',
              'alert_type': 'RECOVER',
              'content': '消息类型: [恢复] 请自己检查问题是否解决,上次报警内容如下\n[2020-01-13 09:05:37]\napplication1最近5分钟内有异常日志1条。最近一条异常信息:\n服务器IP: 127.0.0.1\n异常类型: redis.clients.jedis.exceptions.JedisConnectionException\n自定义信息: HashCache error \n异常信息: save(key:xxxxx, seconds:1200)\n\n详细请看: http://iii50.cn/uMiahM'
            },
            {
              'id': 312,
              'alarm_id': 20,
              'execute_id': 1954,
              'way': 'email',
              'recipient': 'kechangqing',
              'in_silence': 'NO',
              'send_status': 'SUCCESS',
              'create_at': '2020-01-13T09:10:35.000+0800',
              'alert_type': 'RECOVER',
              'content': '消息类型: [恢复] 请自己检查问题是否解决,上次报警内容如下\n[2020-01-13 09:05:37]\napplication1最近5分钟内有异常日志1条。最近一条异常信息:\n服务器IP: 127.0.0.1\n异常类型: redis.clients.jedis.exceptions.JedisConnectionException\n自定义信息: HashCache error \n异常信息: save(key:xxxxxx, seconds:1200)\n\n详细请看: http://iii50.cn/uMiahM'
            },
            {
              'id': 309,
              'alarm_id': 20,
              'execute_id': 1950,
              'way': 'dingding',
              'recipient': 'kechangqing',
              'in_silence': 'NO',
              'send_status': 'SUCCESS',
              'create_at': '2020-01-13T09:05:39.000+0800',
              'alert_type': 'PROBLEM',
              'content': '[2020-01-13 09:05:37]\napplication1最近5分钟内有异常日志1条。最近一条异常信息:\n服务器IP: 127.0.0.1\n异常类型: redis.clients.jedis.exceptions.JedisConnectionException\n自定义信息: HashCache error \n异常信息: save(key:xxxxxx, seconds:1200)\n\n详细请看: http://iii50.cn/uMiahM'
            },
            {
              'id': 310,
              'alarm_id': 20,
              'execute_id': 1950,
              'way': 'email',
              'recipient': 'kechangqing',
              'in_silence': 'NO',
              'send_status': 'SUCCESS',
              'create_at': '2020-01-13T09:05:39.000+0800',
              'alert_type': 'PROBLEM',
              'content': '[2020-01-13 09:05:37]\napplication1最近5分钟内有异常日志1条。最近一条异常信息:\n服务器IP: 127.0.0.1\n异常类型: redis.clients.jedis.exceptions.JedisConnectionException\n自定义信息: HashCache error \n异常信息: save(key:xxxxx, seconds:1200)\n\n详细请看: http://iii50.cn/uMiahM'
            },
            {
              'id': 307,
              'alarm_id': 20,
              'execute_id': 1832,
              'way': 'dingding',
              'recipient': 'kechangqing',
              'in_silence': 'NO',
              'send_status': 'SUCCESS',
              'create_at': '2020-01-13T06:55:36.000+0800',
              'alert_type': 'RECOVER',
              'content': '消息类型: [恢复] 请自己检查问题是否解决,上次报警内容如下\n[2020-01-13 06:50:35]\napplication1最近5分钟内有异常日志1条。最近一条异常信息:\n服务器IP: 127.0.0.1\n异常类型: redis.clients.jedis.exceptions.JedisConnectionException\n自定义信息: HashCache error \n异常信息: save(key:xxxxxxx, seconds:1200)\n\n详细请看: http://iii50.cn/uMiahM'
            },
            {
              'id': 308,
              'alarm_id': 20,
              'execute_id': 1832,
              'way': 'email',
              'recipient': 'kechangqing',
              'in_silence': 'NO',
              'send_status': 'SUCCESS',
              'create_at': '2020-01-13T06:55:36.000+0800',
              'alert_type': 'RECOVER',
              'content': '消息类型: [恢复] 请自己检查问题是否解决,上次报警内容如下\n[2020-01-13 06:50:35]\napplication1最近5分钟内有异常日志1条。最近一条异常信息:\n服务器IP: 127.0.0.1\n异常类型: redis.clients.jedis.exceptions.JedisConnectionException\n自定义信息: HashCache error \n异常信息: save(key:xxxxxx, seconds:1200)\n\n详细请看: http://iii50.cn/uMiahM'
            }
          ]
        }
      }
    }
  }
]
